### DISCUSSION TITLE
Using Vermo with HSM (Hardware security module)

### DISCUSSION THREAD

**Discussion by koptan** [decentralized-identity/veramo#401] [Q&A]
We are planing to use Vermao in our next project ...

But in our system we are saving private keys in HSM device for more security and based on my understanding Veramo will save the Private key in Table.

So my questions are :

1. Can we do such kind of integration between HSM and Veramo ?
2. If yes, then which plugin we need to override? maybe `kms-local` ?
3. Can you show us some example ?

**Comment by koptan** [comment]
one more thing ... our HSM support PKCS11 interface .

**Reply by jasheal** [reply]
Yes, you can write a kms plugin for your HSM. We have an AWS HSM plugin that we hope to open source soon. We will keep this thread posted on its progress.

**Reply by koptan** [reply]
Nice, I will keep waiting.

**Reply by jasheal** [reply]
@koptan I have an update on this. The [Serto](https://github.com/sertoid) team have published [@sertoid/veramo-aws-kms](https://www.npmjs.com/package/@sertoid/veramo-aws-kms) plugin for Veramo on `npm`. The plugin is not open sourced at the moment, but that may change. The Veramo team have not tested this plugin and as it's not part of our core stack, we cannot provide specific support on it.
